src/ImageRenderer.ts

```diff
@@ -294,25 +297,27 @@ export class ImageRenderer implements IDisposable {
     this._renderService = this._terminal._core._renderService;
     this._oldSetRenderer = this._renderService.setRenderer.bind(this._renderService);
     this._renderService.setRenderer = (renderer: any) => {
-      this._removeLayerFromDom();
+      this.removeLayerFromDom();
       this._oldSetRenderer?.call(this._renderService, renderer);
-      this._insertLayerToDom();
     };
-    this._insertLayerToDom();
-    if (this._showPlaceholder) {
-      this._createPlaceHolder();
-    }
   }
 
-  private _insertLayerToDom(): void {
-    this.canvas = ImageRenderer.createCanvas(this._terminal._core._coreBrowserService.window, this.dimensions?.css.canvas.width || 0, this.dimensions?.css.canvas.height || 0);
-    this.canvas.classList.add('xterm-image-layer');
-    this._terminal._core.screenElement?.appendChild(this.canvas);
-    this._ctx = this.canvas.getContext('2d', { alpha: true, desynchronized: true });
+  public insertLayerToDom(): void {
+    if (!this.canvas) {
+      this.canvas = ImageRenderer.createCanvas(this._terminal._core._coreBrowserService.window, this.dimensions?.css.canvas.width || 0, this.dimensions?.css.canvas.height || 0);
+      this.canvas.classList.add('xterm-image-layer');
+      this._terminal._core.screenElement?.appendChild(this.canvas);
```

Comment:
Could `screenElement` not exist here? Should it fail hard instead?

Reply:
Well at some point a headless impl for the addon would be nice - then yes, the screenElement would not be defined here. I initially apified the low level DOM access like in `ImageRenderer.createCanvas` for that reason, to be overloaded later on by a node headless shim. But I did not keep track of that approach in the protocol handlers, which also partially rely on DOM APIs. So to get there more refactoring is needed.

As for the `screenElement` not being defined here - that wont stop the addon from working as long as the canvas could be created. But sure - the `getContext` below should be guarded in case canvas creation failed. Gonna change that and also move the `screenElement` check higher, as there is not much use for a freely dangling canvas element atm.

Reply:
I mean we could assume screenElement exists and throw instead silently "failing". The renderer addons throw on load if it's not on an `open`ed terminal

Reply:
I get your point - yes there is a race condition here, if `onRender` itself is not guarded against the open state. Would you prefer a hard exception in that case, or is a console warning enough (terminal keeps working)?

Reply:
Just found out that `this._coreTerminal._core._coreBrowserService.window` is also undefined for non-opened terminals, which makes the sixel|iip handlers unusable in that state. This was not an issue before, the problem was introduced by this https://github.com/xtermjs/xterm.js/pull/4069 to allow popout terminals. I am currently not sure, if a canvas element can be used for cross document painting or if they'd need to be moved beforehand.

Not gonna address that here, as it needs further investigation (thus there is a currently chance, that images stop working on popout terminals, if the images to be shown originate to the earlier non-popout document)

Reply:
Added an explicit check for `screenElement` with a console warning. Also lifted the `document` association, thus early image loads before calling `.open` should work again.